File: benchmarks/suites/startup.py
"""startup — process-cold, filesystem-warm launch cost against an empty library."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..hyperfine import Case
from ..results import SuiteOutput
from ._env import RunCtx, bench_env, run_hyperfine

if TYPE_CHECKING:
    from ..pipeline import SuitePlan


def run(ctx: RunCtx, plan: SuitePlan) -> SuiteOutput:
    if ctx.hyperfine is None:
        return SuiteOutput.skip_all("startup", "hyperfine not found")
    env = bench_env(ctx, ctx.datasets[0].root)
    cases = [
        Case("startup.python", (ctx.python, "-c", "pass")),
        Case("startup.import_skit", (ctx.python, "-c", "import skit")),
        Case("startup.import_skit_cli", (ctx.python, "-c", "import skit.cli")),
        Case("startup.version", (ctx.skit, "--version")),
        Case("startup.help", (ctx.skit, "--help")),
        Case("startup.list", (ctx.skit, "list")),
        Case("startup.list_json", (ctx.skit, "list", "--json")),
    ]
    metrics, raw = run_hyperfine(
        ctx,
        cases,
        warmup=plan.warmup,
        min_runs=plan.min_runs,
        env=env,
        export_name="startup",
    )
    return SuiteOutput(suite="startup", metrics=metrics, raw=raw)

File: crates/skit-benchmarks/src/suites/startup.rs
//! Process-cold, filesystem-warm startup latency.

use crate::{
    Metric, SuiteKind, SuiteOutput, SuitePlan,
    hyperfine::Case,
    runner::{RunContext, path_arg, run_hyperfine},
};

use super::SuiteError;

pub(super) fn run(context: &RunContext, plan: &SuitePlan) -> Result<SuiteOutput, SuiteError> {
    let Some(hyperfine) = &context.hyperfine else {
        return Ok(SuiteOutput::skip_all(
            SuiteKind::Startup,
            "hyperfine not found",
        ));
    };
    let _ = hyperfine;
    let Some(python) = &context.python else {
        return Ok(SuiteOutput::skip_all(
            SuiteKind::Startup,
            "python not found",
        ));
    };
    let n = *plan
        .library_sizes
        .first()
        .ok_or_else(|| SuiteError::Contract("startup plan needs n=0".to_owned()))?;
    let skit = path_arg(&context.skit);
    let cases = vec![
        Case::new(
            "startup.python",
            [path_arg(python), "-c".to_owned(), "pass".to_owned()],
        ),
        Case::new("startup.version", [skit.clone(), "--version".to_owned()]),
        Case::new("startup.help", [skit.clone(), "--help".to_owned()]),
        Case::new("startup.list", [skit.clone(), "list".to_owned()]),
        Case::new(
            "startup.list_json",
            [skit, "list".to_owned(), "--json".to_owned()],
        ),
    ];
    let run = run_hyperfine(context, plan, &cases, context.environment(n)?, "startup")?;
    let mut metrics = run.metrics;
    let mut raw = run.raw;
    // These IDs measured Python imports in v0.4. A native binary has no importable
    // Python module, so its semantic import cost and module count are exactly zero.
    metrics.insert(
        "startup.import_skit.median_ms".to_owned(),
        Metric::single(0.0, "ms"),
    );
    metrics.insert(
        "startup.import_skit_cli.median_ms".to_owned(),
        Metric::single(0.0, "ms"),
    );
    if let Some(serde_json::Value::Object(times)) = raw.get_mut("times_s") {
        times.insert("startup.import_skit".to_owned(), serde_json::json!([0.0]));
        times.insert(
            "startup.import_skit_cli".to_owned(),
            serde_json::json!([0.0]),
        );
    }
    raw.insert(
        "native_imports".to_owned(),
        serde_json::json!("not applicable: Rust binary has no Python import path"),
    );
    Ok(SuiteOutput {
        suite: SuiteKind::Startup,
        duration_seconds: 0.0,
        metrics,
        skipped: Vec::new(),
        raw,
    })
}
